### PR TITLE
Fixed #28662 -- Silence invalid argument errors in join template-filter.

### PR DESCRIPTION
**Pull request by atombrella**
https://code.djangoproject.com/ticket/28662

### CODE REVIEW

**Comment by timgraham** [comment]
I think `AttributeError`catching can be removed. Please add tests and fix the problem if autoescape is on.

**Comment by atombrella** [comment]
@timgraham I updated it, and added tests. The `if`-branch should also be in the `try-catch`.

File: django/template/defaultfilters.py [hunk 1]
```diff
@@ -517,11 +517,11 @@ def first(value):
...
 def join(value, arg, autoescape=True):
     """Join a list with a string, like Python's ``str.join(list)``."""
-    if autoescape:
-        value = [conditional_escape(v) for v in value]
     try:
+        if autoescape:
+            value = [conditional_escape(v) for v in value]
         data = conditional_escape(arg).join(value)
-    except AttributeError:  # fail silently but nicely
+    except TypeError:  # fail silently but nicely
```
Comment (timgraham) [inline django/template/defaultfilters.py]: I'd update the comment to say `# Fail silently if arg isn't iterable.`

**Comment by atombrella** [comment]
@timgraham Updated. I don't see the harm in tests for both rendering and the function, although they may overlap. I removed the two tests that already pass.

**Pull request merged**